js/src/ipc/writer/binary.ts

```diff
@@ -269,17 +269,17 @@ export class RecordBatchSerializer extends VectorVisitor {
         const byteLength = Math.min(lastOffset - firstOffset, values.byteLength - firstOffset);
         // Push in the order FlatList types read their buffers
         // valueOffsets buffer first
-        this.addBuffer(this.getZeroBasedValueOffsets(offset, length, valueOffsets));
+        this.addBuffer(this.getZeroBasedValueOffsets(0, length, valueOffsets));
         // sliced values buffer second
         this.addBuffer(values.subarray(firstOffset + offset, firstOffset + offset + byteLength));
         return this;
     }
     protected visitListVector<T extends SingleNestedType>(vector: Vector<T>) {
         const { data, length } = vector;
-        const { offset, valueOffsets } = <any> data;
+        const { valueOffsets } = <any> data;
         // If we have valueOffsets (ListVector), push that buffer first
         if (valueOffsets) {
-            this.addBuffer(this.getZeroBasedValueOffsets(offset, length, valueOffsets));
+            this.addBuffer(this.getZeroBasedValueOffsets(0, length, valueOffsets));
         }
         // Then insert the List's values child
         return this.visit((vector as any as ListVector<T>).getChildAt(0)!);
```

Comment:
What went wrong here, exactly? Is the offset being handled elsewhere?

Reply:
@wesm Brian is right, _most_ of the internal buffers are zero-copy sliced up front in the ArrayData internals. This is mostly a performance optimization/complexity reduction, since it allows us to index into those TypedArrays starting from zero in most cases. Notable exceptions are when the indexing is dependent on other buffers (e.g. we _do_ slice the valueOffsets buffer of a Utf8Vector, but not the data buffer), or when multiple values are packed into < 1 byte like the validity bitmap or BoolVector's data buffer. So we still need to track the logical `offset` internally, but only use it in those special cases.

Reply:
@wesm if you want to see a simplified version of the internal buffer management/slicing that's going on here, I have one here in my unfinished [`js-data-refactor`](https://github.com/trxcllnt/arrow/blob/js-data-refactor/js/src/data.ts#L90) branch that I'm reeeally looking forward to getting time to finish soon.

Reply:
Sounds good, looking forward to getting that in :)

Reply:
Yeah, especially since it lets us take full advantage of TypeScript 3's more advanced type inference:

![arrow-inference](https://user-images.githubusercontent.com/178183/46162323-bd9fe880-c23c-11e8-9448-d0fbf3301c02.gif)